app/modules/net.c

```diff
@@ -10,1364 +10,757 @@
 
 #include "c_types.h"
 #include "mem.h"
+#include "osapi.h"
+#include "lwip/err.h"
 #include "lwip/ip_addr.h"
-#include "espconn.h"
 #include "lwip/dns.h" 
+#include "lwip/igmp.h"
+#include "lwip/tcp.h"
+#include "lwip/udp.h"
+
+typedef enum net_type {
+  TYPE_TCP_SERVER = 0,
+  TYPE_TCP_CLIENT,
+  TYPE_UDP_SOCKET
+} net_type;
+
+typedef const char net_table_name[14];
+
+static const net_table_name NET_TABLES[] = {
+  "net.tcpserver",
+  "net.tcpsocket",
+  "net.udpsocket"
+};
+#define NET_TABLE_TCP_SERVER NET_TABLES[0]
+#define NET_TABLE_TCP_CLIENT NET_TABLES[1]
+#define NET_TABLE_UDP_SOCKET NET_TABLES[2]
 
-#define TCP ESPCONN_TCP
-#define UDP ESPCONN_UDP
-
-static ip_addr_t host_ip; // for dns
-
-#ifdef HAVE_SSL_SERVER_CRT
-#include HAVE_SSL_SERVER_CRT
-#else
-__attribute__((section(".servercert.flash"))) unsigned char net_server_cert_area[INTERNAL_FLASH_SECTOR_SIZE];
-#endif
-
-__attribute__((section(".clientcert.flash"))) unsigned char net_client_cert_area[INTERNAL_FLASH_SECTOR_SIZE];
-
-#if 0
-static int expose_array(lua_State* L, char *array, unsigned short len);
-#endif
-
-#define MAX_SOCKET 5
-static int socket_num = 0;
-static int socket[MAX_SOCKET];
-static int tcpserver_cb_connect_ref = LUA_NOREF;  // for tcp server connected callback
-static uint16_t tcp_server_timeover = 30;
-
-static struct espconn *pTcpServer = NULL;
-static struct espconn *pUdpServer = NULL;
+#define TYPE_TCP TYPE_TCP_CLIENT
+#define TYPE_UDP TYPE_UDP_SOCKET
 
-typedef struct lnet_userdata
-{
-  struct espconn *pesp_conn;
+typedef struct lnet_userdata {
+  enum net_type type;
   int self_ref;
-  int cb_connect_ref;
-  int cb_reconnect_ref;
-  int cb_disconnect_ref;
-  int cb_receive_ref;
-  int cb_send_ref;
-  int cb_dns_found_ref;
-#ifdef CLIENT_SSL_ENABLE
-  uint8_t secure;
-#endif
-}lnet_userdata;
+  union {
+    struct tcp_pcb *tcp_pcb;
+    struct udp_pcb *udp_pcb;
+    void *pcb;
+  };
+  union {
+    struct {
+      int cb_accept_ref;
+    } server;
+    struct {
+      int wait_dns;
+      int cb_dns_ref;
+      int cb_receive_ref;
+      int cb_sent_ref;
+      // Only for TCP:
+      int hold;
+      int cb_connect_ref;
+      int cb_disconnect_ref;
+      int cb_reconnect_ref;
+    } client;
+  };
+} lnet_userdata;
+
+#pragma mark - LWIP errors
+
+int luaL_lwip_checkerr (err_t err, lua_State *L) {
```

Comment:
Can I recommend not using the `luaL_` prefix since that's used by the official Lua aux lib? I see what you're going for though, so it's a soft recommendation only :)

Also, customarily the lua_State * is the first arg.